### DISCUSSION TITLE
About strings manipulation

### DISCUSSION THREAD

**Discussion by carlosbazilio** [sampsyo/bril#203] [Q&A] [answered]
Dear,
first of all, congratulations for your excellent work!
I've seen Bril does only deals with integer and boolean values. How about strings? How do you usually deal with it?
Thanks in advance for any hint!
Carlos

**Comment by sampsyo** [comment, marked as answer]
Hi! Good question! The language doesn't have any support for strings, so the answer is mostly "we don't"—people just confine themselves to focusing on numeric programs in a Bril context. If you really want to get creative, you could fake them with the memory extension, which would let you allocate an array of integers representing the characters!

Part of the Bril philosophy is extensibility, so it could be fun to imagine designing a string extension (in the same vein as our current floating-point extension).

**Reply by carlosbazilio** [reply]
Nice! I had thought about it, but I've imagined someone else could've been working on it. I'll get used to Bril to give it a try! ;)

**Reply by carlosbazilio** [reply]
I'd like to know if you have some hints about how to build a Bril extension. For instance, is the extension code for floating point numbers stored at Bril's Github repository? How are Bril's binaries (brili and bril2json, for instance) modified to incorporate these extensions?

**Reply by Pat-Lafon** [reply]
Hello. I would agree that it isn't that clear how to extend Bril as Adrian has noted in https://github.com/sampsyo/bril/issues/38. Because writing an extension is not so well structured, I don't believe there is any "extension code for floating point numbers". Instead, my understanding is that extensions are specified like in https://github.com/sampsyo/bril/tree/main/docs/lang which describe what new fields are used in the Bril JSON by the extension. These extensions are then implemented in tools such as `brili` and `bril2json` as needed. 

For what it's worth, I would start implementing a string extension by laying out what values and operations do you want to use in the extension(are you just sticking with arrays of ints or are you using string literals? do you have chars? Is there a type called "string"? What operations will be supported and are they effect or value operations?). Then you probably want to modify the reference interpreter(`brili`) source code to support this. A quick glance suggests you will need to add to https://github.com/sampsyo/bril/blob/732f63f8548de6e1d5ffdf9f33ce9ed03117f2e5/brili.ts#L103-L141 https://github.com/sampsyo/bril/blob/732f63f8548de6e1d5ffdf9f33ce9ed03117f2e5/brili.ts#L148 https://github.com/sampsyo/bril/blob/732f63f8548de6e1d5ffdf9f33ce9ed03117f2e5/brili.ts#L151-L165 https://github.com/sampsyo/bril/blob/732f63f8548de6e1d5ffdf9f33ce9ed03117f2e5/brili.ts#L167-L180 https://github.com/sampsyo/bril/blob/732f63f8548de6e1d5ffdf9f33ce9ed03117f2e5/brili.ts#L420 https://github.com/sampsyo/bril/blob/732f63f8548de6e1d5ffdf9f33ce9ed03117f2e5/brili.ts#L811-L836
https://github.com/sampsyo/bril/blob/732f63f8548de6e1d5ffdf9f33ce9ed03117f2e5/bril-ts/bril.ts#L13 https://github.com/sampsyo/bril/blob/732f63f8548de6e1d5ffdf9f33ce9ed03117f2e5/bril-ts/bril.ts#L40-L64 https://github.com/sampsyo/bril/blob/732f63f8548de6e1d5ffdf9f33ce9ed03117f2e5/bril-ts/bril.ts#L66-L69

I don't know if I've answered your questions but I hope this post was helpful as I think it would be very interesting to have a String extension in Bril.

**Reply by sampsyo** [reply]
Very well put, @Pat-Lafon! This is a great overview of how to get started with implementing a Bril extension. To summarize, I would personally start with careful thinking about the value types and new opcodes, possibly even by writing the human-language docs first. Then the next step would be to hack the reference interpreter. Eventually, I would probably extend the text syntax, but only as a convenience for the real third step: writing a bunch of tests.
